Fixes various bugs for display order once CPS has merged more fixes

TIHan opened the pull request:
* IMPORTANT *
Merge when CPS pull requests have been also merged. @Pilchie will know which PRs.

This fixes a few bugs with display order, especially for linked files. It also coordinates with CPS fixes for display order too.

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/project-system?pullRequest=3068) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/project-system?pullRequest=3068) before we can accept your contribution.<br/><br/>:x: TIHan [sign now](https://cla.dotnetfoundation.org/dotnet/project-system?pullRequest=3068)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/project-system?pullRequest=3068) it.</sub>

jmarolf commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs:
```diff
@@ -64,6 +66,24 @@ private void ComputeIndices()
+            }
+            
+            return 
+                (propertyContext.Metadata.TryGetValue(FullPathProperty, out var fullPath) && _duplicateOrderMap.TryGetValue(fullPath.Replace(Path.GetDirectoryName(_project.FullPath) + PathSeparator, string.Empty), out index));
```
can we format this to be on multiple lines please?

davkean replied:
This is trying to calculate what exactly? The ItemSpec? ie the value of Include in `<Compile Include="Foo.cs"/>` or the path relative to the project?

The item spec metadata will return the former, UnconfiguredProject.MakeRelative for the later.

jmarolf commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs:
```diff
@@ -15,8 +15,10 @@ namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Order
     {
         private const string FullPathProperty = "FullPath";
+        private const string LinkProperty = "Link";
+        private const string PathSeparator = "\\";
```
Is this guaranteed to always be the path separator character?

davkean commented:
Can we fix up the commits to understand what fixes what? Break up the various bug fixes into separate commits and remove the experimental commits. It makes it easier to rollback if one of the changes causes introduces a regression and helps in the future when we look back and try and understand what changes are made. 

Also explain the *why* in the change - you've spent a bunch of time figuring that out - there's a good reason for them, but you need make it obvious to the reviewer as they may have never looked at this code before, or are haven't spent much time debugging through this code.

davkean commented:
Please add tests for the new cases that you are handling, they are located in src\Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests\ProjectSystem\VS\Tree\Order\TreeItemOrderPropertyProviderTests.cs.

TIHan commented:
Thank you @davkean and @jmarolf for the feedback.

I agree there is code in here that isn't clear what it is doing. I'm marking this as WIP and will make changes according to the feedback.

Though, I do want to explore a better fix for the ordering problem.

TIHan commented:
@davkean I fixed the commits. I still need to add tests as my last task.

The first commit addresses getting ordering working when new changes have been made to the fsproj. This maps to my first PR with CPS that fixes just display ordering.

On second/third commit, the big change here is this: 
Before, when we had duplicate files, we had a separate dictionary that would use their full path as the key instead of their name. However, I found it was a little easier to reason about when using relative paths, so I changed it. The rest of the changes adds ordering support for linked files. This maps to my second PR with CPS.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs:
```diff
@@ -64,27 +75,66 @@ private void ComputeIndices()
+
+            // If we did not get a display order, we need to check its relative path.
+            // If the item has a valid Link property, use its value to find a display order.
+            var isLinked =
```
Use FileItemServices.GetLinkFilePath.

TIHan replied:
Will do.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs:
```diff
@@ -93,6 +143,10 @@ private void ComputeIndices()
                 }
+                else
+                {
+                    propertyValues2.DisplayOrder = 0;
```
These are nodes you don't know - why are setting the display order? Won't this affect the display order of say "Properties"?

TIHan replied:
DisplayOrder is always zero for all items in the beginning, including "Properties". The only time DisplayOrder ever changes is within this provider. Setting display order to zero at this point ensures that everything else stays at zero.

davkean replied:
> The only time DisplayOrder ever changes is within this provider
This may be true right now - but CPS is an extensible system, other teams add nodes to the solution explorer, such as Service Capabilities. We should avoid setting DisplayOrder for nodes we don't own or know about. What issue is this fixing?

TIHan replied:
That is a good point. I guess it's not right to assume. This particular addition isn't fixing anything; so that there may be a red flag to not have this bit. :)

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs:
```diff
@@ -29,11 +33,17 @@ public TreeItemOrderPropertyProvider(IReadOnlyCollection<ProjectItemIdentity> or
 
         public IReadOnlyCollection<ProjectItemIdentity> OrderedItems { get; }
 
+        private static string[] GetStringParts(string evaluatedInclude)
```
"GetPathComponents".

TIHan replied:
Will do.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs:
```diff
@@ -48,13 +58,14 @@ private void ComputeIndices()
                 {
-                    var rootedPath = duplicateFiles.Contains(part) ? _project.MakeRooted(item.EvaluatedInclude) : null;
-                    if (rootedPath != null && !_rootedOrderMap.ContainsKey(rootedPath))
+                    // EvaluatedInclude is a unique relative path for a file.
```
EvaluatedInclude does not have to be relative, it can be rooted; "C:\Foo.cs" or current directory relative "C:Foo.cs".

TIHan replied:
In the second PR for CPS, I ensure that I only give the relative path. This makes sense as I only care about how a file or folder gets rendered relative to the project; that's the path I actually want and the ones we care about ordering. Perhaps, in CPS, ```ProjectItemIdentity``` property "EvaluatedInclude" is a bad name. If it were really "EvaluatedInclude", I would never get the proper link path to do ordering.

TIHan replied:
To note, something like this:
```
<Compile Include="../../ConsoleApp2.sln" />
```
Will actually turn into a linked file, with a link path of just "ConsoleApp2.sln".

davkean replied:
This won't turn in a linked file:
```
<Compile Include="C:\Projects\Project\Program.cs" />
```

When the csproj lives in C:\Projects\Project.

TIHan replied:
It will not, but in my CPS PR, it will return "Program.cs".

davkean replied:
You can't change the identity of an item.

davkean replied:
I think we need to sync up with CPS/offline. You changing some fundamental details of how MSBuild works here. EvaluatedInclude is the identity of an item - you can't make that something else.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs:
```diff
@@ -48,13 +58,14 @@ private void ComputeIndices()
-                    var rootedPath = duplicateFiles.Contains(part) ? _project.MakeRooted(item.EvaluatedInclude) : null;
-                    if (rootedPath != null && !_rootedOrderMap.ContainsKey(rootedPath))
+                    // EvaluatedInclude is a unique relative path for a file.
+                    var duplicatePart = duplicateFiles.Contains(part) ? item.EvaluatedInclude : null;
```
This isn't a duplicatedPart - but rather the path or rootedOrRelativePath of the item.

TIHan replied:
In the previous code, that is what it meant. It doesn't mean that now. 

"duplicatePart" in this case is, for example, like "opt.fs" or "src" in "visualfsharp/src/opt.fs".  Though, we are only checking for duplicate files. In the future, we may check for duplicate folders, but there are some problems with that which are out of scope.

davkean replied:
You are assigning `duplicatedPart` `Item.EvaluatedInclude` which isn't any of those things.

TIHan replied:
I see. I did not explain that well and I was incorrect. ```duplicatePart``` is not any of those and not a good name for this.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs:
```diff
@@ -64,27 +75,66 @@ private void ComputeIndices()
+            {
+                // Link path is always relative.
+                // Make sure we choose the property context with the same name as the last part of the link path.
+                if (string.Equals(propertyContext.ItemName, GetStringParts(linkPath).Last()))
```
What happens with:

```
    <Compile Include="..\Foo.cs">
      <Link>Bar.cs</Link>
    </Compile>
```

TIHan replied:
In this case, CPS will send a property context with an ItemName "Bar.cs" because that is how it is rendered. This same property context will also have a link path that has the same filename as it as well.

However, sometimes CPS will get into a state where it will have both "Foo.cs" and "Bar.cs" property context. This happens when you start changing a file that is not linked, into a linked a file, then back again. When we need to be careful with that because we don't want two items with the same DisplayOrder that is greater than zero.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs:
```diff
@@ -64,27 +75,66 @@ private void ComputeIndices()
+            // Full path needs to turn into a relative path for lookup.
+            return
+                propertyContext.Metadata.TryGetValue(FullPathProperty, out var fullPath) &&
+                _relativeDisplayOrderMap.TryGetValue(_project.MakeRelative(fullPath), out displayOrder);
```
Above this was added as `Item.EvaluatedInclude`, now you've grabbed a normalized path relative to the project. What guarantee do they have to match in something like:

``` XML
<Compile Include="..\Ignored\..\Project\Program.cs" />
```

If you are storing paths in a dictionary for lookup - they have to be normalized.

TIHan replied:
```
<Compile Include="..\Ignored\..\Project\Program.cs" />
```
Becomes a linked file with a path of "Program.cs". So, this is handled.

davkean replied:
What happens to these cases:

```
<Compile Include="Foo\..\Program.cs" />
<Compile Include="Foo\..\\Program.cs" />
<Compile Include="Foo/Program.cs" />
```

TIHan replied:
From the second CPS PR, "Foo\..\Program.cs" and "Foo\..\\Program.cs" would return me "Program.cs".

"Foo/Program.cs" would return me "Foo/Program.cs"

jmarolf commented:
@TIHan can you link to the internal PRs in CPS for continuity's sake?  Thanks!

jmarolf commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs:
```diff
@@ -15,53 +15,105 @@ namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Order
-                    var rootedPath = duplicateFiles.Contains(part) ? _project.MakeRooted(item.EvaluatedInclude) : null;
-                    if (rootedPath != null && !_rootedOrderMap.ContainsKey(rootedPath))
+                    // Folders are special. 
+                    // FIXME: Due to the lack of metadata/info from property context 
```
What changes needs to be made to CPS to get this information?

TIHan replied:
CPS needs to give us a better way of identifying folders by their display/rendering path.

jmarolf commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs:
```diff
@@ -15,53 +15,105 @@ namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Order
+        /// <summary>
+        /// Tries to get a display order for a property context.
+        /// </summary>
+        /// <param name="propertyContext"></param>
```
please either delete these param comments or fill them in.

TIHan commented:
Here are the two PRs on CPS that need to be merged before this can:
https://mseng.visualstudio.com/VSIDEProj/_git/VSIDEProj.CPS/pullrequest/290366
https://mseng.visualstudio.com/VSIDEProj/_git/VSIDEProj.CPS/pullrequest/291520

TIHan commented:
I believe this is finally ready now, if all goes green.

Pilchie commented:
Approved to merge - thanks @TIHan
